src/cc/BPFTable.cc

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2016 Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "BPFTable.h"
+
+#include "bcc_syms.h"
+#include "libbpf.h"
+
+namespace ebpf {
+
+BPFTableBase::BPFTableBase(BPFModule* bpf_module, const std::string& name) {
+  size_t id_ = bpf_module->table_id(name);
+  if (id_ >= bpf_module->num_tables())
+    throw std::invalid_argument("Table " + name + " does not exist");
+  fd_ = bpf_module->table_fd(id_);
+  max_size_ = bpf_module->table_max_entries(id_);
+}
+
+bool BPFTableBase::lookup(void* key, void* value) {
+  return bpf_lookup_elem(fd_, key, value) >= 0;
+}
+
+bool BPFTableBase::next(void* key, void* next_key) {
+  return bpf_get_next_key(fd_, key, next_key) >= 0;
+}
+
+BPFStackTable::~BPFStackTable() {
+  for (auto it : pid_sym_)
+    bcc_free_symcache(it.second, it.first);
+}
+
+std::vector<uint64_t> BPFStackTable::get_stack_addr(int stack_id) {
+  std::vector<uint64_t> res;
+  stacktrace_t stack;
+  if (!lookup(static_cast<void*>(&stack_id), static_cast<void*>(&stack)))
+    return res;
+  int i = 0;
+  while (true) {
+    if ((i == BPF_MAX_STACK_DEPTH) || (stack.ip[i] == 0))
+      break;
+    res.push_back(stack.ip[i]);
+    i++;
+  }
+  return res;
+}
+
+std::vector<std::string> BPFStackTable::get_stack_symbol(
+  int stack_id, int pid
+) {
+  auto addresses = get_stack_addr(stack_id);
+  std::vector<std::string> res;
+  res.reserve(addresses.size());
+
+  if (pid < 0)
+    pid = -1;
+  if (pid_sym_.find(pid) == pid_sym_.end())
+    pid_sym_[pid] = bcc_symcache_new(pid);
+  void* cache = pid_sym_[pid];
+
+  bcc_symbol* symbol = new bcc_symbol();
```

Comment:
I mean, why allocate a `bcc_symbol` with `new` and not just pass a pointer to a local variable allocated on the stack?